brickschema/validate.py

```diff
@@ -0,0 +1,423 @@
+"""
+The `validate` module implements a wrapper of `pySHACL`_ to
+validate an ontology graph against the default Brick Schema constraints (called *shapes*) and user-defined shapes.
+
+.. _`pySHACL`: https://github.com/RDFLib/pySHACL
+"""
+import sys
+import argparse
+import logging
+from rdflib import Graph, Namespace, URIRef, BNode, Literal
+from rdflib.plugins.sparql import prepareQuery
+from .namespaces import BRICK, A, RDF, RDFS, BRICK, BSH, SH, SKOS, bind_prefixes
+import pyshacl
+import io
+import pkgutil
+
+class Validate():
+    """
+    Validates a data graph against Brick Schema and basic SHACL constraints for Brick.  Allows extra
+    constraints specific to the user's ontology.
+    """
+
+    # build accumulative namespace index from participating files
+    # build list of violations, each is a graph
+    def __init__(self, useBrickSchema=True, useDefaultShapes=True):
+        self.log = logging.getLogger()
+        self.log.setLevel(logging.DEBUG if hasattr(sys, '_called_from_test') else logging.WARNING)
+        self.log.info('Validate init')
+
+        self.namespaceDict = {}
+        self.brickG = Graph()
+        self.shapeG = Graph()
+
+        if useBrickSchema:
+            data = pkgutil.get_data(__name__, "ontologies/Brick.ttl").decode()
+            self.brickG.parse(source=io.StringIO(data), format='turtle')
+            self.__buildNamespaceDict(self.brickG)
+
+            # Remove rdfs:domain and rdfs:range.  The modified
+            # ontology will be used for pySHACL reasoning.
+            # See DESIGN.md for more discussion.
+
+            self.brickG.update('DELETE { ?s rdfs:domain ?o .} WHERE { ?s rdfs:domain ?o . }',
+                               initNs=self.namespaceDict)
+            self.brickG.update('DELETE { ?s rdfs:range ?o .} WHERE { ?s rdfs:range ?o . }',
+                               initNs=self.namespaceDict)
+
+        if useDefaultShapes:
+            data = pkgutil.get_data(__name__, "ontologies/BrickShape.ttl").decode()
+            self.shapeG.parse(source=io.StringIO(data), format='turtle')
+            self.__buildNamespaceDict(self.shapeG)
+
+
+    def validate(self, data_graph, shacl_graph=None, ont_graph=None,
+                 inference='rdfs', abort_on_error=False, advanced=True,
+                 meta_shacl=True, debug=False):
+        """
+        Validates data_graph against shacl_graph and ont_graph.
+
+        Args:
+            shacl_graph: default to BrickShape.ttl and shapes added with the addShape[Graph|File]() methods
```

Comment:
The `ont_graph` and `shacl_graph` arguments should either be lists so that they can accept multiple values or, (which I prefer), the Validator should validate an input datagraph using the graphs that have been loaded into the Validator object.

Reply:
The two arguments are laid out to mimic pyshacl.validate() (neither is a list).  I'd like to keep it that way to avoid confusion.  In fact, by offering addShapFile/Graph() we are expanding pyshacl.  I'd like to make the expansion clear to the user.  For pyshacl command, although the user can use multiple -s or -e on the command line, only the last one is taken :(

Reply:
Is there a good reason why we want to preserve the pyshacl API especially when it is broken in this way?

Reply:
There is not a particular bit reason but my conservative habit to preserve an exiting API.  In fact, the brick_validate command takes multiple -s arguments already.  If we use a list, then  the addShapexxx functions can be removed.

Reply:
But why we want multiple ontology graphs?  Multiple shape graphs make more sense.